src/server.py

```diff
@@ -39,13 +42,35 @@ def get_city_data(city_id: str):
     if not (file_path := CONFIG_DIRECTORY_PATH / f"{city_id}.json").is_file():
         raise HTTPException(404, "City not found")
 
-    city_configuration = CityConfiguration.from_path(file_path)
-    city_data_builder = CityDataBuilder(city_configuration)
-
-    return {
-        "tram_track_graph": city_data_builder.tram_track_graph_data,
-        "tram_trips": city_data_builder.tram_trips_data,
-    }
+    if cache.is_valid(city_id):
+        logger.info(f"Loading data from cache for {city_id}")
+        city_data = cache.load(city_id)
+        city_data["last_updated"] = cache.last_updated(city_id)
+        return city_data
+
+    try:
+        city_configuration = CityConfiguration.from_path(file_path)
+        city_data_builder = CityDataBuilder(city_configuration)
+        response_data = {
+            "tram_track_graph": city_data_builder.tram_track_graph_data,
+            "tram_trips": city_data_builder.tram_trips_data,
+            "last_updated": datetime.now().isoformat(),
+        }
+        cache.save(city_id, response_data)
+        return response_data
+
+    except Exception as exc:
+        logger.exception(f"Failed to build city data for {city_id}: {exc}")
```

Comment:
When logging an exception, you can pass the `exc` as a keyword argument in order to provide it with more context:
```suggestion
        logger.exception(f"Failed to build city data for {city_id}", exc_info=exc)
```